src/daemon/daemon.cpp

```diff
@@ -1129,14 +1129,33 @@ InstanceSnapshotsMap map_snapshots_to_instances(const InstanceSnapshotPairs& ins
 }
 
 bool verify_snapshot_picks(const InstanceSelectionReport& report,
-                           const std::unordered_map<std::string, SnapshotPick>& snapshot_picks)
+                           const std::unordered_map<std::string, SnapshotPick>& snapshot_picks,
+                           bool purge)
 {
     auto any_snapshot = false;
+    std::vector<std::string> snapshots_of_deleted_instances{};
     for (const auto* selection : {&report.deleted_selection, &report.operative_selection})
```

Comment:
Basically, `snapshots_of_deleted_instances` and the `if (selection == &report.deleted_selection && (!pick_it->second.all_or_none || !purge))` and the throwing later are only for `report.deleted_selection`. If we split the loop into two code blocks, then it will be more clear and have more localized variables. However, that causes more code duplication. I guess it is a trade-off and we can leave it like this.

Reply:
Yes, the rationale is indeed less code duplication. If I had two blocks, you would still need to read them to understand that they are mostly the same. This makes common code immediate and still separates the difference in that inner `if`. It is not ideal, but it seems to me to be the "least bad" of the two options.